alibi_detect/od/backend/torch/base.py

```diff
@@ -0,0 +1,196 @@
+from __future__ import annotations
+from typing import List, Dict, Union, Optional
+from dataclasses import dataclass, asdict
+from abc import ABC, abstractmethod
+
+import numpy as np
+import torch
+
+from alibi_detect.od.backend.torch.ensemble import FitMixinTorch
+
+
+@dataclass
+class TorchOutlierDetectorOutput:
+    """Output of the outlier detector."""
+    threshold_inferred: bool
+    scores: torch.Tensor
+    threshold: Optional[torch.Tensor]
+    preds: Optional[torch.Tensor]
+    p_vals: Optional[torch.Tensor]
+
+    def numpy(self) -> Dict[str, Union[bool, Optional[torch.Tensor]]]:
+        """Converts the output to numpy."""
+        outputs = asdict(self)
+        for key, value in outputs.items():
+            if isinstance(value, torch.Tensor):
+                outputs[key] = value.cpu().detach().numpy()
+        return outputs
+
+
+class TorchOutlierDetector(torch.nn.Module, FitMixinTorch, ABC):
+    """Base class for torch backend outlier detection algorithms."""
+    threshold_inferred = False
+    threshold = None
+
+    def __init__(self):
+        super().__init__()
+
+    @abstractmethod
+    def _fit(self, x_ref: torch.Tensor) -> None:
+        """Fit the outlier detector to the reference data.
+
+        Parameters
+        ----------
+        x_ref
+            Reference data.
+
+        Raises
+        ------
+        `NotImplementedError`
+            Raised if not implemented.
+        """
+        raise NotImplementedError()
+
+    @abstractmethod
+    def score(self, X: torch.Tensor) -> torch.Tensor:
+        """Score the data.
+
+        Parameters
+        ----------
+        X
+            Data to score.
+
+        Raises
+        ------
+        `NotImplementedError`
+            Raised if not implemented.
+        """
+        raise NotImplementedError()
+
+    @torch.jit.ignore
```

Comment:
Yeah so torchscript doesn't like control flow so we have to use one of [`@torch.jit.ignore`](https://pytorch.org/docs/stable/generated/torch.jit.ignore.html) or [`@torch.jit.unused`](https://pytorch.org/docs/stable/generated/torch.jit.unused.html). Here I was actually using the wrong one namely `ignore` which doesn't allow you to export scripted models. I've now changed it to `unused`. Note `unused` decorated function or method raises an error if it's incorrectly scripted. So to avoid this we add a check that we're scripting. For example:

```py
    ...
    def transform(self, x: torch.Tensor):
        if not torch.jit.is_scripting():
            self.check_fitted()
        return self._transform(x)
```

This is just for models that we want to torchscript and export to then deploy in that format.

Reply:
Just to highlight, the implication of this is presumably that `NotFitException`'s will not be raised by detectors that have been torch scripted, i.e. by detectors in deployment. Don't think this is something to worry about too much for now, but wonder if we should do one or both of the following:

1. Ensure this caveat is highlighted in the documentation somewhere, as I imagine users will at some point deploy unfitted detectors and get unclear error messages.
2. Open an issue for us to think about possible solutions for this, or at least keep an eye on future `torch` versions to see if they improve this torch script limitation.

Reply:
Added an issue #761